src/algebra/lie_algebra.lean

```diff
@@ -0,0 +1,214 @@
+/-
+Copyright (c) 2019 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import ring_theory.algebra data.matrix.basic
+
+/-!
+# Lie algebras
+
+This file defines Lie rings, and Lie algebras over a commutative ring. It shows how these arise from
+associative rings and algebras via the ring commutator. In particular it defines the Lie algebra
+of endomorphisms of a module as well as of the algebra of square matrices over a commutative ring.
+
+## Notations
+
+We introduce the notation ⁅x, y⁆ for the Lie bracket. Note that these are the Unicode "square with
+quill" brackets rather than the usual square brackets.
+
+## Implementation notes
+
+Lie algebras are defined as modules with a compatible Lie ring structure, and thus are partially
+unbundled. Since they extend Lie rings, these are also partially unbundled.
+
+## References
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 1--3*][bourbaki1975]
+
+## Tags
+
+lie bracket, ring commutator, jacobi identity, lie ring, lie algebra
+-/
+
+universes u v
+
+/--
+A product operation, intended use in Lie algebras and similar structures.
+-/
+class has_bracket (L : Type v) := (bracket : L → L → L)
+
+notation `⁅`x`,` y`⁆` := has_bracket.bracket x y
+
+namespace ring_commutator
+
+variables {A : Type v} [ring A]
+
+/--
+The ring commutator captures the extent to which a ring is commutative. It is identically zero
+exactly when the ring is commutative.
+-/
+def commutator (x y : A) := x*y - y*x
+
+local notation `⁅`x`,` y`⁆` := commutator x y
+
+@[simp] lemma add_left (x y z : A) :
+  ⁅x + y, z⁆ = ⁅x, z⁆ + ⁅y, z⁆ :=
+by { unfold commutator, simp [right_distrib, left_distrib] }
```

Comment:
Do you know you could write `by simp [commutator, right_distrib, left_distrib]`? The same trick applies elsewhere of course.